src/Mobs/Path.cpp

```diff
@@ -0,0 +1,391 @@
+
+		// Calculate H. This is A*'s Heuristics value.
+		#if DISTANCE_MANHATTAN == 1
+			// Manhattan distance. DeltaX + DeltaY + DeltaZ.
+			a_Cell->m_H = 10 * (abs(a_Cell->m_Location.x-m_Destination.x) + abs(a_Cell->m_Location.y-m_Destination.y) + abs(a_Cell->m_Location.z-m_Destination.z));
+		#else
+			// Euclidian distance. sqrt(DeltaX^2 + DeltaY^2 + DeltaZ^2), more precise.
+			a_Cell->m_H = std::sqrt((a_Cell->m_Location.x - m_Destination.x) * (a_Cell->m_Location.x - m_Destination.x) * 100 + (a_Cell->m_Location.y - m_Destination.y) * (a_Cell->m_Location.y - m_Destination.y) * 100 + (a_Cell->m_Location.z - m_Destination.z) * (a_Cell->m_Location.z - m_Destination.z) * 100);
+		#endif
+
+		#if HEURISTICS_ONLY == 1
+			a_Cell->m_F = a_Cell->m_H;  // Greedy search. https://en.wikipedia.org/wiki/Greedy_search
+		#else
+			a_Cell->m_F = a_Cell->m_H + a_Cell->m_G;  // Regular A*.
+		#endif
+
+		OpenListAdd(a_Cell);
+		return;
+	}
+
+	// Case 3: Cell is in the open list, check if G and H need an update.
+	int NewG = a_Caller->m_G + a_GDelta;
+	if (NewG < a_Cell->m_G)
+	{
+		a_Cell->m_G = NewG;
+		a_Cell->m_H = a_Cell->m_F + a_Cell->m_G;
+		a_Cell->m_Parent = a_Caller;
+	}
+
+}
+
+
+
+
+
+cPathCell * cPath::GetCell(const Vector3d & a_Location)
+{
+	// Create the cell in the hash table if it's not already there.
+	cPathCell * Cell;
+	if (m_Map.count(a_Location) == 0)  // Case 1: Cell is not on any list. We've never checked this cell before.
+	{
+		Cell = new cPathCell();
+		Cell->m_Location = a_Location;
+		m_Map[a_Location] = Cell;
+		Cell->m_IsSolid = IsSolid(a_Location);
+		Cell->m_Status = eCellStatus::NOLIST;
+		#ifdef COMPILING_PATHFIND_DEBUGGER
+			#ifdef COMPILING_PATHFIND_DEBUGGER_MARK_UNCHECKED
+				si::setBlock(a_Location.x, a_Location.y, a_Location.z, debug_unchecked, Cell->m_IsSolid ? NORMAL : MINI);
+			#endif
+		#endif
+		return Cell;
+	}
+	else
+	{
+		return m_Map[a_Location];
+	}
+}
+
+
+
+
+
+// Add the next point in the final path.
+void cPath::AddPoint(Vector3d a_Vector)
+{
+	m_PathPoints.push_back(a_Vector);
+	++m_PointCount;
+}
+
+
+
+
+
+#ifndef COMPILING_PATHFIND_DEBUGGER
+bool cPath::Item(cChunk * a_Chunk)  // returns FALSE if there's a solid or if we failed.
+{
+	int RelX = m_Item_CurrentBlock.x - a_Chunk->GetPosX() * cChunkDef::Width;
+	int RelZ = m_Item_CurrentBlock.z - a_Chunk->GetPosZ() * cChunkDef::Width;
+
+	// If it's setmode, Sets a block to cobblestone and exits, will be removed eventually.
+	if (m_Item_SetMode)
+	{
+		a_Chunk->FastSetBlock(RelX, m_Item_CurrentBlock.y, RelZ, E_BLOCK_COBBLESTONE, 0);
+		return false;
+	}
+
+
+	if (!a_Chunk->IsValid())
+	{
+		return false;
+	}
+	BLOCKTYPE BlockType;
+	NIBBLETYPE BlockMeta;
+	a_Chunk->GetBlockTypeMeta(RelX, m_Item_CurrentBlock.y, RelZ, BlockType, BlockMeta);
+	return (!cBlockInfo::IsSolid(BlockType));
+
+	// TODO Maybe I should queue several blocks and call item() at once for all of them for better performance?
+	// I think Worktycho said each item() call needs 2 locks.
```

Comment:
Yes, much better indeed. But is this performance critical or can I do it later? doWithChunk will be called at most 100 times per 2 seconds for each mob, and typically much less than that. If it's critical, I'll do it now, otherwise, I'll add it to my to do list.